### PR TITLE
GH-39134: [Java] Create module info compiler plugin

### LINKED ISSUE #39134: [Java] (JPMS) Add a maven plugin for compiling modules while retaining JDK8 support
### Describe the enhancement requested

The traditional method of compiling modules is to use -release 9 to compile everything in the module, then recompile everything except module-info.java using release 8. This conflicts with the use of Unsafe in JDK8.

Instead, we can write a plugin which does a purely syntactic compilation of module-info.java files.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by jduo**
### Rationale for this change
Create and integrate a maven plugin for compiling module-info.java files without using maven-compiler-plugin
and release 9+. This is necessary for supporting Java 8 unsafe code, which has different than unsafe code
in JDK 9.

### What changes are included in this PR?
- Add the module-info-compiler-maven-plugin and utilize it in the codebase.
- Exclude module-info.java files from maven-compiler-plugin.

### Are these changes tested?
Yes, they generate valid module-info.class files in JARs and test JARs.

### Are there any user-facing changes?
No
* Closes: #39134

### CODE REVIEW

File: java/maven/module-info-compiler-maven-plugin/pom.xml [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    <dependency>
+      <groupId>org.apache.maven.plugin-tools</groupId>
+      <artifactId>maven-plugin-annotations</artifactId>
+      <version>3.6.0</version>
+      <scope>provided</scope>
+    </dependency>
+  </dependencies>
+
+  <build>
+    <pluginManagement><!-- lock down plugins versions to avoid using Maven defaults (may be moved to parent pom) -->
```
Comment (lidavidm) [inline java/maven/module-info-compiler-maven-plugin/pom.xml:75]: Why do we want to avoid the defaults/are these inconsistent with the global config?

**Review by lidavidm — APPROVED** [review]

File: java/maven/module-info-compiler-maven-plugin/pom.xml (hunk 1 above)
Comment (jduo) [reply java/maven/module-info-compiler-maven-plugin/pom.xml:75]: This was generated by the maven plugin archetype I used.

**Comment by lidavidm** [comment]
@github-actions crossbow submit *java*

**Comment by danepitkin** [comment]
dependabot is trying to update some of the dependencies added here. Should we ignore them? How do we determine the right support matrix for dependencies?

https://github.com/apache/arrow/pull/39202
https://github.com/apache/arrow/pull/39201

**Comment by kou** [comment]
@jduo It seems that this broke our documentation CI job. Could you check it?

https://github.com/apache/arrow/actions/runs/7182308293/job/19558525786#step:7:13088

```text
Error:  Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project arrow-maven-plugins: Execution default-site of goal org.apache.maven.plugins:maven-site-plugin:3.3:site failed: A required class was missing while executing org.apache.maven.plugins:maven-site-plugin:3.3:site: org/apache/maven/doxia/siterenderer/DocumentContent
Error:  -----------------------------------------------------
Error:  realm =    plugin>org.apache.maven.plugins:maven-site-plugin:3.3
Error:  strategy = org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy
Error:  urls[0] = file:/root/.m2/repository/org/apache/maven/plugins/maven-site-plugin/3.3/maven-site-plugin-3.3.jar
Error:  urls[1] = file:/root/.m2/repository/org/apache/maven/reporting/maven-reporting-exec/1.1/maven-reporting-exec-1.1.jar
Error:  urls[2] = file:/root/.m2/repository/org/apache/maven/reporting/maven-reporting-api/3.0/maven-reporting-api-3.0.jar
Error:  urls[3] = file:/root/.m2/repository/org/apache/maven/shared/maven-shared-utils/0.3/maven-shared-utils-0.3.jar
Error:  urls[4] = file:/root/.m2/repository/com/google/code/findbugs/jsr305/2.0.1/jsr305-2.0.1.jar
Error:  urls[5] = file:/root/.m2/repository/org/codehaus/plexus/plexus-component-annotations/1.5.5/plexus-component-annotations-1.5.5.jar
Error:  urls[6] = file:/root/.m2/repository/org/sonatype/aether/aether-util/1.7/aether-util-1.7.jar
Error:  urls[7] = file:/root/.m2/repository/org/eclipse/aether/aether-util/0.9.0.M2/aether-util-0.9.0.M2.jar
Error:  urls[8] = file:/root/.m2/repository/org/sonatype/sisu/sisu-inject-bean/1.4.2/sisu-inject-bean-1.4.2.jar
Error:  urls[9] = file:/root/.m2/repository/org/sonatype/sisu/sisu-guice/2.1.7/sisu-guice-2.1.7-noaop.jar
Error:  urls[10] = file:/root/.m2/repository/org/codehaus/plexus/plexus-interpolation/1.14/plexus-interpolation-1.14.jar
Error:  urls[11] = file:/root/.m2/repository/org/sonatype/plexus/plexus-sec-dispatcher/1.3/plexus-sec-dispatcher-1.3.jar
Error:  urls[12] = file:/root/.m2/repository/org/sonatype/plexus/plexus-cipher/1.4/plexus-cipher-1.4.jar
Error:  urls[13] = file:/root/.m2/repository/org/apache/maven/maven-archiver/2.4.2/maven-archiver-2.4.2.jar
Error:  urls[14] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-sink-api/1.4/doxia-sink-api-1.4.jar
Error:  urls[15] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-logging-api/1.4/doxia-logging-api-1.4.jar
Error:  urls[16] = file:/root/.m2/repository/junit/junit/3.8.1/junit-3.8.1.jar
Error:  urls[17] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-core/1.4/doxia-core-1.4.jar
Error:  urls[18] = file:/root/.m2/repository/xerces/xercesImpl/2.9.1/xercesImpl-2.9.1.jar
Error:  urls[19] = file:/root/.m2/repository/xml-apis/xml-apis/1.3.04/xml-apis-1.3.04.jar
Error:  urls[20] = file:/root/.m2/repository/org/apache/httpcomponents/httpclient/4.0.2/httpclient-4.0.2.jar
Error:  urls[21] = file:/root/.m2/repository/commons-logging/commons-logging/1.1.1/commons-logging-1.1.1.jar
Error:  urls[22] = file:/root/.m2/repository/commons-codec/commons-codec/1.3/commons-codec-1.3.jar
Error:  urls[23] = file:/root/.m2/repository/org/apache/httpcomponents/httpcore/4.0.1/httpcore-4.0.1.jar
Error:  urls[24] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-module-xhtml/1.4/doxia-module-xhtml-1.4.jar
Error:  urls[25] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-module-apt/1.4/doxia-module-apt-1.4.jar
Error:  urls[26] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-module-xdoc/1.4/doxia-module-xdoc-1.4.jar
Error:  urls[27] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-module-fml/1.4/doxia-module-fml-1.4.jar
Error:  urls[28] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-module-markdown/1.4/doxia-module-markdown-1.4.jar
Error:  urls[29] = file:/root/.m2/repository/org/pegdown/pegdown/1.2.1/pegdown-1.2.1.jar
Error:  urls[30] = file:/root/.m2/repository/org/parboiled/parboiled-java/1.1.4/parboiled-java-1.1.4.jar
Error:  urls[31] = file:/root/.m2/repository/org/parboiled/parboiled-core/1.1.4/parboiled-core-1.1.4.jar
Error:  urls[32] = file:/root/.m2/repository/org/ow2/asm/asm/4.1/asm-4.1.jar
Error:  urls[33] = file:/root/.m2/repository/org/ow2/asm/asm-tree/4.1/asm-tree-4.1.jar
Error:  urls[34] = file:/root/.m2/repository/org/ow2/asm/asm-analysis/4.1/asm-analysis-4.1.jar
Error:  urls[35] = file:/root/.m2/repository/org/ow2/asm/asm-util/4.1/asm-util-4.1.jar
Error:  urls[36] = file:/root/.m2/repository/javax/servlet/servlet-api/2.5/servlet-api-2.5.jar
Error:  urls[37] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-decoration-model/1.4/doxia-decoration-model-1.4.jar
Error:  urls[38] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-site-renderer/1.4/doxia-site-renderer-1.4.jar
Error:  urls[39] = file:/root/.m2/repository/org/apache/velocity/velocity-tools/2.0/velocity-tools-2.0.jar
Error:  urls[40] = file:/root/.m2/repository/commons-beanutils/commons-beanutils/1.7.0/commons-beanutils-1.7.0.jar
Error:  urls[41] = file:/root/.m2/repository/commons-digester/commons-digester/1.8/commons-digester-1.8.jar
Error:  urls[42] = file:/root/.m2/repository/commons-chain/commons-chain/1.1/commons-chain-1.1.jar
Error:  urls[43] = file:/root/.m2/repository/commons-validator/commons-validator/1.3.1/commons-validator-1.3.1.jar
Error:  urls[44] = file:/root/.m2/repository/dom4j/dom4j/1.1/dom4j-1.1.jar
Error:  urls[45] = file:/root/.m2/repository/sslext/sslext/1.2-0/sslext-1.2-0.jar
Error:  urls[46] = file:/root/.m2/repository/org/apache/struts/struts-core/1.3.8/struts-core-1.3.8.jar
Error:  urls[47] = file:/root/.m2/repository/antlr/antlr/2.7.2/antlr-2.7.2.jar
Error:  urls[48] = file:/root/.m2/repository/org/apache/struts/struts-taglib/1.3.8/struts-taglib-1.3.8.jar
Error:  urls[49] = file:/root/.m2/repository/org/apache/struts/struts-tiles/1.3.8/struts-tiles-1.3.8.jar
Error:  urls[50] = file:/root/.m2/repository/commons-collections/commons-collections/3.2.1/commons-collections-3.2.1.jar
Error:  urls[51] = file:/root/.m2/repository/org/apache/maven/doxia/doxia-integration-tools/1.5/doxia-integration-tools-1.5.jar
Error:  urls[52] = file:/root/.m2/repository/org/codehaus/plexus/plexus-archiver/1.0/plexus-archiver-1.0.jar
Error:  urls[53] = file:/root/.m2/repository/org/codehaus/plexus/plexus-io/1.0/plexus-io-1.0.jar
Error:  urls[54] = file:/root/.m2/repository/org/codehaus/plexus/plexus-i18n/1.0-beta-7/plexus-i18n-1.0-beta-7.jar
Error:  urls[55] = file:/root/.m2/repository/org/apache/velocity/velocity/1.5/velocity-1.5.jar
Error:  urls[56] = file:/root/.m2/repository/oro/oro/2.0.8/oro-2.0.8.jar
Error:  urls[57] = file:/root/.m2/repository/org/codehaus/plexus/plexus-velocity/1.1.8/plexus-velocity-1.1.8.jar
Error:  urls[58] = file:/root/.m2/repository/org/codehaus/plexus/plexus-utils/1.5.10/plexus-utils-1.5.10.jar
Error:  urls[59] = file:/root/.m2/repository/org/mortbay/jetty/jetty/6.1.25/jetty-6.1.25.jar
Error:  urls[60] = file:/root/.m2/repository/org/mortbay/jetty/servlet-api/2.5-20081211/servlet-api-2.5-20081211.jar
Error:  urls[61] = file:/root/.m2/repository/org/mortbay/jetty/jetty-util/6.1.25/jetty-util-6.1.25.jar
Error:  urls[62] = file:/root/.m2/repository/commons-lang/commons-lang/2.5/commons-lang-2.5.jar
Error:  urls[63] = file:/root/.m2/repository/commons-io/commons-io/1.4/commons-io-1.4.jar
Error:  Number of foreign imports: 1
Error:  import: Entry[import  from realm ClassRealm[maven.api, parent: null]]
Error:  
Error:  -----------------------------------------------------: org.apache.maven.doxia.siterenderer.DocumentContent
```

**Pull request merged**